txp/Pos/Txp/Toil/Logic.hs

```diff
@@ -87,33 +88,38 @@ rollbackToil txun = do
 processTx ::
        (HasTxpConfiguration, HasConfiguration)
     => BlockVersionData
+    -> Set Address
     -> EpochIndex
     -> (TxId, TxAux)
     -> ExceptT ToilVerFailure LocalToilM TxUndo
-processTx bvd curEpoch tx@(id, aux) = do
+processTx bvd blacklist curEpoch tx@(id, aux) = do
     whenM (lift $ hasTx id) $ throwError ToilKnown
     whenM ((>= memPoolLimitTx) <$> lift memPoolSize) $
         throwError (ToilOverwhelmed memPoolLimitTx)
-    undo <- mapExceptT utxoMToLocalToilM $ verifyAndApplyTx bvd curEpoch True tx
+    undo <- mapExceptT utxoMToLocalToilM $ verifyAndApplyTx bvd blacklist curEpoch True tx
     undo <$ lift (putTxWithUndo id aux undo)
 
 -- | Get rid of invalid transactions.
 -- All valid transactions will be added to mem pool and applied to utxo.
 normalizeToil ::
        (HasTxpConfiguration, HasConfiguration)
     => BlockVersionData
+    -> Set Address
     -> EpochIndex
     -> [(TxId, TxAux)]
     -> LocalToilM ()
-normalizeToil bvd curEpoch txs = mapM_ normalize ordered
+normalizeToil bvd blacklist curEpoch txs = mapM_ normalize ordered
   where
+    -- If there is a cycle in the tx list, topsortTxs returns Nothing.
+    -- Why is that not an error? And if its not an error, why bother
+    -- top-sorting them anyway?
```

Comment:
:thinking:

Reply:
Is it possibly a performance improvement to be topologically sorted if possible, and not a correctness issue?

Reply:
If that's the case, a comment would have helped, right?

I keep finding things in this code where there are important assumptions that are undocumented or things like this where there's something strange with zero explanation.

Reply:
Yeah I'm just throwing out wild guesses